presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -226,4 +227,155 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
 
         checkArgument((totalBitLength <= maximumBits), format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
     }
+
+    /**
+     * Searches for and outputs the addresses within certain ranges in each dimension.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of addresses within the given value ranges
+     */
+    public List<Long> zOrderSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        List<Long> addressRanges = zOrderCondensedSearchCurve(minimumValues, maximumValues);
+
+        List<Long> addresses = new ArrayList<>();
+        for (int index = 0; index < addressRanges.size(); index += 2) {
+            long minAddress = addressRanges.get(index);
+            long maxAddress = addressRanges.get(index + 1);
+            for (long address = minAddress; address <= maxAddress; address++) {
+                addresses.add(address);
+            }
+        }
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    // TODO: Create own "pair" object
```

Comment:
I was about to comment the same and then saw this TODO :) Let's address this? I think using `List<Range>` would be much clearer.